### ISSUE TITLE
lot of redirections

### ISSUE THREAD

**Issue by mblendinger** [engintron/engintron#1232]
using latest version of cpanel and engintron, I can't use it, when I enable engintron, websites start failing about redirections.
some tip please ?

**Comment by mblendinger** [comment]
note: the problem is with websites using CloudFlare, try changing settings, like ssl settings,but I cant found a silver bullet

**Comment by fevangelou** [comment]
Did you check the Engintron docs? https://engintron.com/docs/#/pages/Engintron-&-CloudFlare

**Comment by mblendinger** [comment]
> 
> 
> Did you check the Engintron docs? https://engintron.com/docs/#/pages/Engintron-&-CloudFlare

of course, but thank you.

**Comment by fevangelou** [comment]
Nginx does not perform any redirects.

If you use Flexible SSL on CloudFlare and have .htaccess redirects from HTTP to HTTPS what do you expect will happen?

**Comment by faca5** [comment]
Hello.

We have similar problem with latest version of Engintron for cPanel/WHM - v1.13.0 on subdomain. On previous version works without problem. Before upgrade works normally.

Settings:
CloudFlare OFF
.htaccess redirect OFF

Problem is ONLY with MAIN domain without "index.php".

Example:
https://example.com/		not working, redirecting
https://example.com/index.php	working
https://example.com:8443/	working

Error:
The page isn’t redirecting properly

**Comment by fevangelou** [comment]
Then you have some internal PHP redirect which is not properly checking the current request protocol and blindly redirects to HTTPS either way.

If it's a WordPress site, some plugin is most likely to blame.

**Comment by Serjudo** [comment]
Hello, I have the same problem or a very similar one.

I am testing Enginetron on my test server, before moving it to production, but I have redirection problems that I have not been able to solve so far.

In my case, the testing domain does not have Cloudfare enabled. If I deactivate Enginetron it works fine. If I restart the Nginx and Apache services, it starts working after a while but without touching anything it suddenly starts to fail.

When it's failing, if I directly access the root domain https://desarrollopapelstore.factoriadigitalpremium.es/ it doesn't work, if instead of the root domain, I access an internal page of the domain it works without problems, and access to the root domain starts working (I guess because of a cookie issue). If I close the window and try again, it fails again.

Any suggestions?

**Comment by fevangelou** [comment]
@Serjudo There is most likely some broken /index.php to / redirect in the site's .htaccess file.

**Comment by faca5** [comment]
Old version works without problem.

Look like first visit - is the problem.
We have "cookie" checking.

It is problem only on /

**Comment by fevangelou** [comment]
@faca5 Can you show me your code? A URL would also help me test your sent HTTP headers...

**Comment by faca5** [comment]
Url:
https://alert.studiofaca.com/

.htaccess has:
RewriteCond %{HTTPS} off
RewriteRule .* https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]

If you visit 
https://alert.studiofaca.com/
it will show
The page isn’t redirecting properly

An error occurred during a connection to alert.studiofaca.com.

    This problem can sometimes be caused by disabling or refusing to accept cookies.

If you visit
https://alert.studiofaca.com/index.php
and then
https://alert.studiofaca.com/
it will show content.

On previous version have havent all these problem.

If we FLUSH cache then for ~24 hours will work
https://alert.studiofaca.com/

It is very strange. Looks like first visit create wrong cache and then nginx retrive wrong cache.
We have upgraded on latest version but we will downgrade if this error still persists.

**Comment by fevangelou** [comment]
@faca5 The important difference between v1.12 and v1.13 is that Engintron's Nginx setup is now configured for better caching. If you send the right HTTP headers, Nginx will properly cache the content (server-side).

Test your site with: `curl -I "https://alert.studiofaca.com/"`

Notice that:
a) you still get a 301 redirect, which means your .htaccess redirect is bad. Remove that .htaccess rule, purge Engintron's cache and check the URL using a private window.
b) you have an extremely high cache-control header (`cache-control: max-age=259200`). Nginx respects that and serves a cached copy. Cookies are ignored when building its cache.
c) Now try querying the domain using a random query string, something that an uptime tool would use for example. E.g. https://alert.studiofaca.com/?t=20201002 - You'll now get a 406 response (!). This indicates the app is not routing all requests properly.

Moreover, this app seems to be entirely for user-generated content. As such, you should 100% disable any caching for it, using Engintron's custom rules.

And to conclude, the issues are entirely app specific.

My guess is that similar issues occur for everyone else.

@all Guys, please do some digging on your own too. Use curl to examine your affected site's headers. Triple-check your .htaccess rules (don't just use the first one that's returned on a Google result).

**Comment by fevangelou** [comment]
@faca5 Bonus point:
```
$ curl -I "https://alert.studiofaca.com/index.php"
HTTP/2 406 
server: nginx
date: Sat, 10 Oct 2020 09:10:25 GMT
content-type: text/html; charset=iso-8859-1
content-length: 373
x-xss-protection: 1; mode=block
x-content-type-options: nosniff
```

Routing is wrong in the app (returns a 406 HTTP code). Nginx only caches responses returned with a 200 code. Thus, no caching occurs. That's why you see the updated page.

Both issues you have (redirect & caching) are all because of bad .htaccess rules and bad app routing.

**Comment by Serjudo** [comment]
Thank you for the answers,

I can confirm that my problems seem to have been solved, eliminating all the redirections that I had configured in the .htaccess file. Both those that transformed the requests to https, and those that removed or added www before the url.

I have also activated it in production (domain with cloudfare enabled) and after removing the redirects it also seems to work perfectly.

@fevangelou  Maybe you should add an entry in the documentation, talking about this, because it seems that at least in my case, this kind of redirects are not compatible with Engintron.

Thanks for your help.

**Comment by fevangelou** [comment]
@Serjudo It's not "just" redirects. It's bad redirects :) Apache .htaccess rules that *may* work in Apache (which is more forgiving compared to Nginx) does not mean they'll work in Nginx too when exposed to it in a proxy scenario.

**Comment by fevangelou** [comment]
Besides, I can't possibly document that "hey, if you got bad .htaccess rules, things will break". This should be common sense.

**Comment by faca5** [comment]
Hello.

We will check.  Thank you.

**Comment by Serjudo** [comment]
@fevangelou I agree that these are possibly bad redirections, I am certainly not an expert on this subject.

If it helps, these are the redirections that previously worked with Apache, and after the installation of Engintron, they don't work properly:

For the https redirection, I used this first:
  RewriteCond %{HTTPS} off
  RewriteRule .* https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]

and then I tried this:
  RewriteCond %{HTTP:X-Forwarded-Proto} !https
  RewriteRule ^(.*)$ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]

this for the redirection, adding www at the beginning:
  RewriteCond %{HTTP_HOST} ^xxxxxxxxxx.es
  RewriteCond %{HTTP_HOST} !^xxxxxx.xxxxxxxxxx.es
  RewriteRule ^(.*)$ https://www.xxxxxxxxxx.es/$1 [R=301,L]

**Comment by Serjudo** [comment]
Hello again,

I'm afraid I have problems again from time to time with redirection, which only affect access to the root domain of the page and are solved by clearing the cache in nginx.

It was working perfectly for more than 3 days and without touching anything in the configuration, such problems recur from time to time.

In production I have Cloudfare activated but only for DNS redirection. The SSL/TLS encryption mode is Full (I have a self-signed SSL certificate installed on the server) and I added in the custom rules of Nginx
set $PROXY_DOMAIN_OR_IP "xxx.xxx.xx.xx" as indicated in the docs.

On Cloudfare I also have the following options enabled:
Always use HTTPS
Automatic HTTPS rewrites

As for the .htaccess file, I deactivated the redirections I thought were presenting the problem. These are the ones that I still have active, that working in Apache work perfectly:

  RewriteCond %{REQUEST_URI} "/testimonial/"
  RewriteRule (.*) $1 [L]
  RewriteCond %{REQUEST_URI} "/downloader/"
  RewriteRule (.*) $1 [L]
  RewriteCond %{REQUEST_URI} "/rss/"
  RewriteRule (.*) $1 [L]

  ## rewrite API2 calls to api.php (by now it is REST only)
  RewriteCond %{REQUEST_URI} !^/[0-9]+\..+\.cpaneldcv$
  RewriteCond %{REQUEST_URI} !^/\.well-known/pki-validation/[A-F0-9]{32}\.txt(?:\ Comodo\ DCV)?$
  RewriteRule ^api/rest api.php?type=rest [QSA,L]

  ## TRACE and TRACK HTTP methods disabled to prevent XSS attacks
  RewriteCond %{REQUEST_METHOD} ^TRAC[EK]
  RewriteCond %{REQUEST_URI} !^/[0-9]+\..+\.cpaneldcv$
  RewriteCond %{REQUEST_URI} !^/\.well-known/pki-validation/[A-F0-9]{32}\.txt(?:\ Comodo\ DCV)?$
  RewriteRule .* - [L,R=405]

  RewriteCond %{REQUEST_URI} !(downloader) [NC]
  RewriteCond %{THE_REQUEST} ^.*/index.php
  RewriteRule ^(.*)index.php$ https://www.papelstore.es/$1 [R=301,L] 

## always send 404 on missing files in these folders
  RewriteCond %{REQUEST_URI} !^/(media|skin|js)/
  
## exception for mod_pagespeed URLs
  RewriteCond %{REQUEST_URI} ! pagespeed

  RewriteCond %{REQUEST_FILENAME} !-f
  RewriteCond %{REQUEST_FILENAME} !-d
  RewriteCond %{REQUEST_FILENAME} !-l

  ## rewrite everything else to index.php
  RewriteRule .* index.php [L]

  ## BLOCK git & svn and redirect to homepage access to dir var
  RewriteRule .*\.git/.* - [F]
  RewriteRule [F]
  RewriteRule ^var/(.*)$ www.papelstore.es [R=301,NC,L]

Can you help me?

**Comment by jackwakefield** [comment]
Adding `$scheme` back to `proxy_cache_key` solved this for us.

**Comment by fevangelou** [comment]
@jackwakefield Adding $scheme to https://github.com/engintron/engintron/blob/master/nginx/proxy_params_dynamic#L78 will simply create 2 cached copies of the same content. Not really efficient, unless you serve different content indeed.

@Serjudo I've demonstrated already and in detail that bad redirections and/or bad HTTP headers coming from a site/app are the cause of such issues. I can't possibly do this forever. Engintron is a free plugin. Do some (thorough) digging on your own.

**Comment by faca5** [comment]
> 
> 
> Adding `$scheme` back to `proxy_cache_key` solved this for us.

Hello.

It works.

We had redirect from HTTP to HTTPS.

Thank you very much.

**Comment by fevangelou** [comment]
I'm curious to see that redirect of yours @faca5 which causes such a weird issue...

**Comment by fedealvz** [comment]
Thank you @fevangelou 

When .htaccess have two redirects (1st for "http to https" and 2nd for "non www" to "www") it will redirect the first rule, but the second it will cause "too_many_redirects".

```
# 1st rule "http to https"
RewriteCond %{HTTPS} !on
RewriteRule ^(.*)$ https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]

# 2nd rule "non www to www"
RewriteCond %{HTTP_HOST} !^www\. [NC]
RewriteRule .* https://www.%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
```

Response:

```
http://domain.com
	301 Moved Permanently
https://domain.com/
	301 Moved Permanently
https://domain.com/
	301 Moved Permanently
https://domain.com/
	301 Moved Permanently
...etc...
```

**Comment by fevangelou** [comment]
I think cPanel's redirect is better. See this:
```
RewriteCond %{HTTPS} !=on [OR]
RewriteCond %{HTTP:X-Forwarded-Proto} =http
RewriteRule .* https://%{HTTP_HOST}%{REQUEST_URI} [L,R=301]
```

In any case, I've just released v1.14.0 which re-instates the $scheme in proxy_cache_key. This should resolve your redirect issues.
